vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java

```diff
@@ -1659,4 +1659,65 @@ public int commitCount() throws IOException {
     public Hash initialHash() {
         return EMPTY_TREE;
     }
+
+    @Override
+    public Optional<Hash> wholeHash(String rev) {
+        try (var p = capture("git", "rev-parse", rev)) {
+            var res = p.await();
+            if (res.status() == 0 && res.stdout().size() == 1) {
+                return Optional.of(new Hash(res.stdout().get(0)));
+            }
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public Optional<List<String>> stagedFile(Path path) {
+        try (var p = capture("git", "cat-file", "-p", ":" + path.toString())) {
+            var res = p.await();
+            if (res.status() == 0) {
+                return Optional.of(res.stdout());
+            }
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Commit staged() throws IOException {
+        var author = new Author("jcheck", "jcheck@none.none");
+        var commitMetaData = new CommitMetadata(new Hash("staged"), List.of(head()), author, ZonedDateTime.now(),
+                author, ZonedDateTime.now(), List.of(""));
+        return new Commit(commitMetaData, List.of(diffStaged()));
+    }
+
+    @Override
+    public Commit workingTree() throws IOException {
+        var author = new Author("jcheck", "jcheck@none.none");
+        var commitMetaData = new CommitMetadata(new Hash("working-tree"), List.of(head()), author, ZonedDateTime.now(),
+                author, ZonedDateTime.now(), List.of(""));
+        return new Commit(commitMetaData, List.of(diff(head())));
+    }
```

Comment:
I tried to implement case 5 by Erik's idea. It seems work fine. But I don't know whether my implementation is acceptable, like the author and hash in the fake commit.

Reply:
I'd be more worried about what it might do to my current index or working tree. Is there any chance things could go badly and mess that up?

Reply:
> I'd be more worried about what it might do to my current index or working tree. Is there any chance things could go badly and mess that up?

No, I believe under any condition it won't do anything to your index or working tree.

Reply:
> I'd be more worried about what it might do to my current index or working tree. Is there any chance things could go badly and mess that up?

It will only create a fake `Commit` object but never use git to commit

Reply:
Zhao is correct, this is completely read-only in regards to the local Git repository.